libinteract/libinteract.py

```diff
@@ -867,6 +867,116 @@ def calc_cg_fullmatrix(identifiers, idxs, percmat, perco):
+            else:
+                output.append(output_chain)
+        # Add all nodes that are in contact with different chains
+        logical_vector = array_T[0] != array_T[sec_res]
+        output_diff_chain = array[logical_vector]
+        if output_diff_chain.shape[0] == 0:
+            warnstr = "No interchain contacts found" 
+            log.warning(warnstr)
+        else:
+            output.append(output_diff_chain)
+    return output
+
+def create_matrix_list(fullmatrix, table_list, pdb, hb = False):
+    """Takes in the full matrix and returns a list of matrices. The
+    first matrix is the full matrix while the following matrices contain
+    intrachain persistences (if present). The final matrix contains 
+    interchain persistences (if present).
+    """
+
+    mat_list = [fullmatrix]
+    # Select which column has the resid
+    if hb:
+        sec_res_id = 5
+    else:
+        sec_res_id = 4
+    # If contacts exist in multiple chains
+    if len(table_list) > 1:
+        # Map each residue id to a position in the matrix
+        resids = pdb.residues.resids
+        res_dict = {str(resids[i]):i for i in range(fullmatrix.shape[0])}
+        # For all tables excluding the first one
+        for i in range(1, len(table_list)):
+            # Create empty matrix
+            matrix = np.zeros(fullmatrix.shape)
+            # Get list of indexes for the matrix
+            table = table_list[i].T
+            mat_i = [res_dict[i] for i in table[1]]
+            mat_j = [res_dict[j] for j in table[sec_res_id]]
+            # Fill the matrix with appropriate values
+            matrix[mat_i, mat_j] = fullmatrix[mat_i, mat_j]
+            matrix[mat_j, mat_i] = fullmatrix[mat_j, mat_i]
+            mat_list.append(matrix)
+    return mat_list
+
+
+def save_output_list(table_list, filename, mat_list=None, hb=False):
+    """Save each element in the list as a separate file. The first
+    element contains the whole dataset, followed by intrachain
+    data (if present). The final element contains interchain data.
+    """
+
+    # Check if the table or the matrix is being saved and change the parameters
+    if mat_list is not None:
+        ext = ".dat"
+        data = mat_list
+        delim = ' '
+        format = "%.1f"
+    else:
+        ext = ".csv"
+        data = table_list
+        delim = ','
+        format = '%s'
+    for i in range(len(table_list)):
+        # For each element, change the filename and save the file
+        if i == 0:
+            fname = filename + "_all_chains" + ext
+        else:
+            # Check which column has the second residue
+            if hb:
+                sec_res = 4
+            else:
+                sec_res = 3
+            chain1 = table_list[i][0][0]
+            chain2 = table_list[i][0][sec_res]
+            # Determine if intra or interchain
+            if chain1 == chain2:
+                fname = filename + "_intra_chain_" + str(chain1) + ext
+            else:
```

Comment:
If you have more than 2 chains in your system, by not specifying which chains you are calculating the intra-chain contacts between this file gets overwritten.

Reply:
I'm not sure I understand? The print function takes in a list of arrays and prints a file for every element in the list. So if I have 5 arrays in the list, it will save 5 different files.  The chain1 and chain2 variables only look in the 1st and the 3rd column of the array to check the name of the chain.

Reply:
I'm not sure if this is the correct way to test, but I changed the chain name for some of the atoms in the test pdb file so that it had a chain A, B, C, D and E. There were no overwritten files, there were 5 output files and each had different content.

Reply:
My bad, I wrote intra-chain but actually meant inter-chains, sorry! It seems like you do not specify the chains in the output name (for inter-chains interactions).

Reply:
Ohhhhh, wait then there is a mistake then. The interchain output is basically all the chains e.g. A-B, B-C, etc. I thought all interchain outputs were supposed to be together, but I guess this means I should split them as well?

Reply:
No, it is one output per pair since also the corresponding matrix with only the interactions between that specific pair of chains must be written separately :) basically one txt and one matrix per each chain (for intra-chain interactions) and per each pair of chains (for inter-chain interactions).

Reply:
I'm working on a fix right now

Reply:
Perfect! Thank you